src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts

```diff
@@ -1,87 +1,135 @@
+    }
+
+    const firstNonZeroByte = this.findFirstNonZeroByteIndex();
+
+    // bufferValue is empty, or contains only 0 bytes
+    if (firstNonZeroByte === this.bufferValue.length) {
+      return Quantity.ZERO_VALUE_STRING;
     }
+
+    let value = this.bufferValue.toString("hex", firstNonZeroByte);
+
+    // only need to check the first char, as we have already skipped 0 bytes in call to this.bufferValue.toString().
+    if (value[0] === "0") {
+      value = value.slice(1);
+    }
+
+    return `0x${value}`;
   }
+
   public toBuffer(): Buffer {
-    // 0x0, 0x00, 0x000, etc should return BUFFER_EMPTY
-    if (Buffer.isBuffer(this.value)) {
-      // trim zeros from start
-      let best = 0;
-      for (best = 0; best < this.value.length; best++) {
-        if (this.value[best] !== 0) break;
-      }
-      if (best > 0) {
-        return this.value.slice(best);
-      } else {
-        return this.value;
-      }
-    } else if (typeof this.value === "string") {
-      let val = this.value.slice(2).replace(/^(?:0+(.+?))?$/, "$1");
-      if (val === "" || val === "0") {
-        return BUFFER_EMPTY;
-      }
-    } else if (this.value === 0 || this.value === 0n) {
+    if (this.bufferValue == null) {
       return BUFFER_EMPTY;
     }
 
-    return super.toBuffer();
-  }
-  public toBigInt(): bigint | null {
-    const value = this.value;
+    const firstNonZeroByte = this.findFirstNonZeroByteIndex();
 
-    // TODO(perf): memoize this stuff
-    if (Buffer.isBuffer(value)) {
-      const bigInt = bufferToBigInt(value);
-      return bigInt == null ? (this._nullable ? null : 0n) : bigInt;
+    if (firstNonZeroByte > 0) {
+      return this.bufferValue.subarray(firstNonZeroByte);
     } else {
-      return value == null ? (this._nullable ? null : 0n) : BigInt(value);
+      return this.bufferValue;
     }
   }
+
+  public toBigInt(): bigint | null {
+    if (this.bufferValue == null) {
+      return this._nullable ? null : 0n;
+    }
+    if (this.bufferValue.length === 0) {
+      return 0n;
+    }
+    return bufferToBigInt(this.bufferValue);
+  }
+
   public toNumber() {
-    // TODO(perf): memoize this stuff
-    return typeof this.value === "number"
-      ? this.value
-      : Number(this.toBigInt());
+    if (this.bufferValue == null) {
+      return this._nullable ? null : 0;
+    }
+
+    const firstNonZeroByte = this.findFirstNonZeroByteIndex();
+
+    const length = this.bufferValue.length - firstNonZeroByte;
+    if (length === 0) {
+      return 0;
+    }
+
+    let result: number;
+    if (length > 6) {
+      const trimmedBuffer = firstNonZeroByte === 0 ? this.bufferValue : this.bufferValue.subarray(firstNonZeroByte, length);
+      result = Number(bufferToBigInt(trimmedBuffer));
+
+      if (!Number.isSafeInteger(result)) {
+        console.warn(`0x${this.bufferValue.toString("hex")} is too large - the maximum safe integer value is 0${Number.MAX_SAFE_INTEGER.toString(16)}`);
```

Comment:
I see the warning but I do not see where `result` is being set to `Number.MAX_SAFE_INTEGER.toString(16)` or that value is being returned instead.

Reply:
@tenthirtyone the error message states that  the given value is too largebecausee the maximum _safe_ integer value is MAX_SAFE_INTEGER. It is not implying the maximum value is MAX_SAFE_INTEGER.

Reply:
We return whatever value is parsed, with the warning indicating that the returned value might start to lose precision.